### ISSUE TITLE
Docs: Secrets for vault doesn't note that the "key" requested is literally "value"

### ISSUE THREAD

**Issue by d3rp** [buildbot/buildbot#5017]
For context: here's an excerpt from the buildbot source that is responsible for querying a secret from the hashicorp vault:
`            ret = json.get('data', {}).get('data', {}).get('value')`

When reading vault documentation, the version data section is not emphasized leading to believe this is a typical key value store i.o.w. one has to land on the specific section covering the kv secrets engine to find out there's more to it. Took me a while to realize that in vault you create a secret which maps as the key in buildbot and insert in that secret in vault a versioned data with `key="value", value=<actual secret>` which is then retrieved by buildbot with the aforementioned snippet in `vault.py`. Would be helpful to have this explained in the docs.

Edit:
Actually, while reviewing https://learn.hashicorp.com/vault/secrets-management/sm-versioned-kv#web-ui-2 I'm puzzled as to why is that hardcoded "value" used. Could the `master/buildbot/secrets/providers/vault.py` file have a key field available along the lines of:
```
    @defer.inlineCallbacks
    def get(self, entry):
        """
        get the value from vault secret backend
        """
        key = "value"
        if '/' in entry:
            parts = entry.split('/')
            if len(parts) > 2:
                raise KeyError("The key %s has to have only 1 or 2 parts separated with '/'" % entry)
            entry = parts[0]
            key = parts[1]

        if self.apiVersion == 1:
            path = self.secretsmount + '/' + entry
        else:
            path = self.secretsmount + '/data/' + entry

        # note that the HTTP path contains v1 for both versions of the key-value
        # secret engine. Different versions of the key-value engine are
        # effectively separate secret engines in vault, with the same base HTTP
        # API, but with different paths within it.
        proj = yield self._http.get('/v1/{0}'.format(path))
        code = yield proj.code
        if code != 200:
            raise KeyError("The key %s does not exist in Vault provider: request"
                           " return code:%d." % (entry, code))
        json = yield proj.json()
        if self.apiVersion == 1:
            ret = json.get('data', {}).get(key)
        else:
            ret = json.get('data', {}).get('data', {}).get(key)
        return ret
```

**Comment by d3rp** [comment]
Actually, while reviewing https://learn.hashicorp.com/vault/secrets-management/sm-versioned-kv#web-ui-2 I'm puzzled as to why is that hardcoded "value" used. Could the `master/buildbot/secrets/providers/vault.py` file have a key field available along the lines of:
```
    @defer.inlineCallbacks
    def get(self, entry):
        """
        get the value from vault secret backend
        """
        key = "value"
        if '/' in entry:
            parts = entry.split('/')
            if len(parts) > 2:
                raise KeyError("The key %s has to have only 1 or 2 parts separated with '/'" % entry)
            entry = parts[0]
            key = parts[1]

        if self.apiVersion == 1:
            path = self.secretsmount + '/' + entry
        else:
            path = self.secretsmount + '/data/' + entry

        # note that the HTTP path contains v1 for both versions of the key-value
        # secret engine. Different versions of the key-value engine are
        # effectively separate secret engines in vault, with the same base HTTP
        # API, but with different paths within it.
        proj = yield self._http.get('/v1/{0}'.format(path))
        code = yield proj.code
        if code != 200:
            raise KeyError("The key %s does not exist in Vault provider: request"
                           " return code:%d." % (entry, code))
        json = yield proj.json()
        if self.apiVersion == 1:
            ret = json.get('data', {}).get(key)
        else:
            ret = json.get('data', {}).get('data', {}).get(key)
        return ret
```

**Comment by phillxnet** [comment]
> Took me a while to realise that in vault you create a secret which maps as the key in buildbot and insert in that secret in vault a versioned data with `key="value", value=<actual secret>`

@d3rp Yes me to. I agree that more clarification would be good there but of course it would be better to improve buildbots flexibility code wise anyway, which is what you seem to suggest. I'd looked at the buildbot code to find these assumptions only after reading up on Vault myself. To me it looks to have been coded, in buildbot, around the default arrangement found within the official Vault docker image. These are not necessarily representative of typical use in my opinion but do server to get Buildbot up and running quickly with the docker Vault instantiation. Even though that is indicated, within the Vault docs, to be at least suspect for production use: i.e. un-sealted I believe (from my memory anyway).

My only contribution, thus far, to buildbot was a minor improvement (hopefully) to this manual entry via pr:
"Vault secret management doc additions" #4937

And I was tempted to emphasise the key as literal "value" requirement further but was also pushed for time. I also though that it was at least in embryonic form in the example given at the very end of the page. From my recent limited experience in this project I'd guess they would welcome further pull requests if they 'pass muster' so it would I assume be appreciated by all if you fancy having a go. Security is hard enough as is and I was personally chuffed to find Vault support in Buildbot but as you surmise it is still a little limited. I'm not personally able to dive into this one just yet on the code side but I'd encourage you to contribute to one or both of docs/code as others will surely have the same difficulties we have both had on this.

Apologies for my prior meagre doc effort here not quite making things clear enough for the current state of the code, but I am also a new comer to both Vault and Buildbot so though baby steps would be best first.

**Comment by d3rp** [comment]
I'm new to vault too. In the end I got vault working for a PoC so the docs were good in that respect. It's tricky to write exhaustive docs and keep within scope of the project and time frames. It felt like the vault almost requires a short example of setting the whole combination up.

**Comment by tardyp** [comment]
@d3rp Now that it is fresh for you, do you want to try yourself at it?

**Comment by d3rp** [comment]
I don't have the time for that.

**Comment by p12tic** [comment]
This has been fixed in #6058 and in a some previous updating the documentation.
